### ISSUE TITLE
Support to override macros

### ISSUE THREAD

**Issue by rahulsom** [asciidoctor/asciidoctor#2801]
I would like to modify the `include` macro to call the original `include` macro with one small change, add a link to the github url to edit the file.  

Is that currently possible? 
Would that make sense for asciidoctor to have something like that?

**Comment by Mogztter** [comment]
You could also add a preprocessor to add the edit link after the `include` macro. Basically this extension will convert:

```adoc
include::https://raw.githubusercontent.com/asciidoctor/asciidoctor/master/README.adoc[]
```

into:

```adoc
include::https://raw.githubusercontent.com/asciidoctor/asciidoctor/master/README.adoc[]

https://github.com/asciidoctor/asciidoctor/edit/master/README.adoc[Edit on Github]
```

**Comment by mojavelinux** [comment]
> to call the original include macro with one small change

This is the part that's not well supported. It's possible if you dive into internal APIs, but that's not ideal.

I think we definitely need to rethink the use cases here so that it's possible to accomplish minor, medium, and large customizations to the include processing.

One thing that will help with the minor is #571, which would allow to change how the path is resolved, but not have to deal with the processing. If that were to allow you to return a File object, then you could even modify the input stream before Asciidoctor reads it.

But there might be other cases were you want to be able to delegate to super. Perhaps this is as simple as making the default process method of the IncludeProcessor perform the same work as the built-in processor.

**Comment by Mogztter** [comment]
> But there might be other cases were you want to be able to delegate to super. Perhaps this is as simple as making the default process method of the IncludeProcessor perform the same work as the built-in processor.

Indeed, I have another use case where I want Asciidoctor to resolve the include target, handle tag filtering and line filtering but then apply some changes to the content (before it gets included).

**Comment by mojavelinux** [comment]
I keep running into cases where I need it too. I would esp like to get rid of the duplicate logic in Antora. I think we should consider this a key goal for an early minor release of 2, perhaps even 2.1 if we can swing it.
